pkg/standalone-utils/contracts/PermitRelayer.sol

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/IERC20Permit.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/IERC20PermitDAI.sol";
+import "@balancer-labs/v2-vault/contracts/interfaces/IVault.sol";
+import "./base/MultiDelegatecall.sol";
+
+/**
+ * @title PermitRelayer
+ * @notice Allows users to atomically perform multiple Balancer Vault actions in sequence
+ * including token approvals using permit (where supported)
+ */
+contract PermitRelayer is MultiDelegatecall {
+    IVault public immutable vault;
+
+    constructor(IVault _vault) {
+        vault = _vault;
+    }
+
+    /**
+     * @notice Allows calling an arbitrary function on the Vault
+     */
+    function _vaultAction(uint256 value, bytes memory data) private returns (bytes memory) {
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool success, bytes memory result) = address(vault).call{ value: value }(data);
+
+        // Pass up revert if the call failed
+        if (!success) {
+            // solhint-disable-next-line no-inline-assembly
+            assembly {
+                returndatacopy(0, 0, returndatasize())
+                revert(0, returndatasize())
+            }
+        }
+
+        return result;
+    }
+
+    function setRelayerApproval(
+        address relayer,
+        bool approved,
+        bytes calldata authorisation
+    ) external payable {
+        bytes memory data = abi.encodePacked(
```

Comment:
```suggestion
        require(relayer == address(this));
        bytes memory data = abi.encodePacked(
```

Not sure about this, what do you think?

Reply:
Quite probably worth it. It prevents a malicious frontend opting the user in to a relayer without their knowledge by including it in a multicall and it's unlikely that a user will want to approve a batch of relayers without directly interacting with them at the same time.

We should probably have a `approveRelayerAndCall` function which activates the relayer and then delegatecalls into another of its functions. This could then be standard on all relayers.

Reply:
Can then decide whether we want to use `approveRelayerAndCall` on this relayer as it's an unnecessary abstraction on top of a multicall.

Reply:
Addressed in #671